staking/effective/calculate.go

```diff
@@ -111,16 +113,31 @@ func Compute(
 		if slotsCount == 0 {
 			continue
 		}
+		shardSlotsCount := make([]int, shardCount)
 		spread := numeric.NewDecFromBigInt(staker.slot.Stake).
 			QuoInt64(int64(slotsCount))
+		startIndex := len(eposedSlots)
 		for i := 0; i < slotsCount; i++ {
-			eposedSlots = append(eposedSlots, SlotPurchase{
+			slot := SlotPurchase{
 				Addr: staker.addr,
 				Key:  staker.slot.SpreadAmong[i],
 				// NOTE these are same because later the .EPoSStake mutated
 				RawStake:  spread,
 				EPoSStake: spread,
-			})
+			}
+			shard := new(big.Int).Mod(slot.Key.Big(), big.NewInt(int64(shardCount))).Int64()
+			shardSlotsCount[int(shard)]++
+			if slotsLimit > 0 && shardSlotsCount[int(shard)] > slotsLimit {
+				continue
+			}
+			eposedSlots = append(eposedSlots, slot)
+		}
+		if effectiveSlotsCount := len(eposedSlots) - startIndex; effectiveSlotsCount != slotsCount {
```

Comment:
Please rename "effective" to "limited" to be more clear (e.g. limitedSlotsCount, limitedSpread etc.). Because these logic are for limitedSlot, it's not for the original EffectivePoS design.